### ISSUE TITLE
map_min_of_outlinks_to_node and similar: handling of non-existent links

### ISSUE THREAD

**Issue by gregtucker** [landlab/landlab#205]
In raster_mappers.py, there are a series of functions that seek the max or min of link values for each node. In our current data structures, a "missing" link is flagged as a link ID of -1. If you fetch a link value corresponding to this, you will actually get the link value for the last link in the array, rather than skipping it, which probably should be the correct behavior.

In map_min_of_outlinks_to_node, I changed the doctest values to what I *think* they should be. @mcflugen, can you have a look?

**Comment by SiccarPoint** [comment]
LONG: SORRY

Random idea, feel free to shoot me down…

Almost all actual data is stored in fields, right? Because this is a data structure *we* designed, we have control over it. It would be quite straightforward to make it so that under the hood (and not exposed to the user) all fields are actually nnodes+1 long, and the final value is a zero. When you ask for or input the values, LL could just hide this zero from you. Then all this will “just work”.

I’m not sure if this is more confusing to the user than it’s worth, but it could be pretty cool. So:
```
>>> mg.at_node[‘myvals’][mg.node_neighbors].sum(axis=1)
```
…always gives the expected behavior, at high speed, with efficient numpy operations, with no faffing about. We could also potentially override the indexing operator to have an “assert vals[-1] == 0” statement, to minimise the chances of bad scenarios if the user supplies a non-field array.

Really just spitballing here, but what do people think? Can certainly see a case this is just confusing.


@gregtucker: the problem with “just skipping it” is that this isn’t really possible. You have to choose between:

1. break the efficient numpy operations by using some kind of jagged list (VERY expensive),
2. put in, e.g., BAD_INDEX_VALUE such that if the user can’t do something stupid, but then have to use a (expensive) search-and-replace on the BAD_INDEX_VALUE before you can use the array in *any* use, or
3. what we have: use -1 and trust the user to have RTFM and know this will give unexpected results if they’re not careful.

AFAIK, those are the only three options. You seem to be advocating (2), but note this is still expensive, and the user basically ends up wasting a bunch of processor time just to get the array back to condition (3) anyway.

A compromise might be to have these functions carry an input argument, bad_ID=BAD_INDEX_VALUE. This is nice, as it means a user who hasn’t read the docs ends up with the idiot-proof condition (2), but a more savvy user can specify “-1” and recover the nice behaviour of (3). I had something like this set up on one function somewhere, and it seemed to work nicely, with a 2-item dict under the hood, with the keys “-1” and “BAD_INDEX_VALUE”. The downside is that this uses twice as much memory, for what’s already a fairly chunky array (iff the user is using both arguments in different places).

Could save memory alternatively by only storing the one with -1s, then pasting in BAD_INDEX_VALUE every time it gets called with the default (I suggest this way round, as someone specifying the -1 option is clearly more concerned with speed than the other putative user).

**Comment by SiccarPoint** [comment]
Just realised I may have interpreted your comment wrong, sorry. You’re not advocating not using -1s at all are you? Oops. Ignore me. Though the first idea may be worthwhile anyway.

**Comment by mcflugen** [comment]
It certainly would be easy to add an extra element when we store fields. I don't see a problem with that.

However, a potential problem would be that we would have to create copies of data. At least in these sorts of cases,
```python
z = np.arange(12)
grid = RasterModelGrid((3, 4))
grid.at_node['myvals'] = z
```
I have *z* being created by `numpy.arange` but perhaps, instead, it comes from another component or read in from a file. In this case (I think), the expectation of the user is that they are storing a reference to their data - not a copy.

**Comment by SiccarPoint** [comment]
Fair point. Though I guess they’re giving us, e.g., z as an input, so under the hood we could do something like

```
>>> field = z.append(0.)
>>> z = field[:-1]
```

So we copy the data once, but then reset *their* variable to *our* field. Possibly this wouldn’t work, but maybe…?

**Comment by mcflugen** [comment]
We could definitely do some trickery to make it look like we didn't create a copy. For instance we could change the data buffer of their array to point to another buffer (one element longer), which we allocate, but we still would have to copy their data to the new buffer.

It's possible that numpy arrays are given a little extra space in their buffers for cases as this. If so, no problem, we wouldn't have to make a copy.

**Comment by mcflugen** [comment]
As an example,
```python
>>> x = np.arange(4, dtype=float)
>>> y = np.empty(5, dtype=float)
>>> y[:-1] = x[:]
>>> y[-1] = 0.
>>> x.data = y.data
>>> y[0] = 99
>>> x
array([ 99.,   1.,   2.,   3.])
>>> x.size
4
>>> y.size
5
```
Although it appears that *x* didn't change, and it's linked to *y*, we still have to make a copy. I don't think there's anyway around it. As such, there's a big performance hit but perhaps it's worth it?

**Comment by mcflugen** [comment]
Let's do some benchmarks. @SiccarPoint, could you give some sample code that demonstrates all of this? Maybe something like?

The set-up,
```python
>>> n_nodes = 1000000
>>> vals_at_node = np.arange(n_nodes, dtype=float)
>>> nodes = np.random.random_integers(0, n_nodes, 1000)
>>> bad_nodes = np.random.random_integers(0, len(nodes), 200)
>>> nodes[bad_nodes] = -1
```
and then this
```python
>>> our_vals_at_node = np.empty(n_nodes + 1, dtype=float)
>>> our_vals_at_node[-1] = 0.
>>> our_vals_at_node[:-1] = vals_at_node
>>> vals_at_node.data = our_vals_at_node.data
>>> our_vals_at_node[nodes].sum()
```
versus this
```python
>>> nodes = nodes[nodes >= 0]
>>> vals_at_node[nodes].sum()
```
The performance will depend on `n_nodes` and the number of bad values, I suppose. So maybe there's a cross-over when one method is faster than the other. I have to say, faster or not, the first way does look nicer (remember those first 4 lines are executed only one within the fields class so usually you wouldn't see them). The second also requires a copy when removing the bad nodes from the `nodes` array  and this copy has to be done every time. In the first case, the first 4 lines are only done once and, in fact, if the field is created with `zeros`, `ones`, `empty`, `add_zeros` etc. it wouldn't have to be done at all.

Another potential problem is that, depending on the operation and the quantity, the bad value may not always be 0. What if the operation is `min` or `max` instead of `sum`? However, since this would all be hidden to the user (well, depending on which "user" we're talking about), we could just set the bad value to whatever the operation calls for.

**Comment by SiccarPoint** [comment]
I don’t think the second case works for lots of LL cases, which is the problem. The issue is that most arrays that use the -1s are 2d. IIRC, the line

>>> nodes = nodes[nodes >= 0]

ends up flattening the array, which means you lose the ability to sum the nodes round each point. Which is exactly why you need to be able to do something involving -1s, so you can at least rely on actions like:

vals_incl_badvals = vals[neighbors_with_-1s].sum(axis=1)
vals_good_in_core = vals_incl_badvals[mg.core_nodes]

…then work the edges independently, if you need to, or just use the 0-at-end trick on the fly. I don’t actually think there’s any other easy way. You could loop through the lines, but that’s a terrible idea.

Short answer - your proposed test I don’t think is representative of the majority of LL use cases. I would anticipate in your case it would be a close run thing, but in a realistic 2d LL case much slower to not use a trailing zero.

**Comment by mcflugen** [comment]
Can you give me a better test, then? What do you think the fastest way to do the second method is?

My performance tests are showing method 1 is slightly faster calculating the sums. However, because of the big array copy in that method, the total time really depends on the size of the array. However, for typical landlab-sized arrays it's not all that much of a hit. For instance, for a 1000000 node grid, the copy takes about .004 seconds and is linear with the number of nodes after that.

**Comment by SiccarPoint** [comment]
As I alluded to in my original comment, I can’t think of a way that doesn’t boil down to “take some actions to get it like version 1, then do that”, which is - obviously - slower. That’s what we’re doing in all current uses.

I think the issue boils down to: is the robustness of setting this up so it works automatically work it for the potential confusion it will cause for users?

**Comment by mcflugen** [comment]
I think it may not be clear, so just in case, I'm leaning toward method 1 (for those that haven't been following along, that's the one with the extra element at the end).

For a `1000 x 1000` node grid, the second method using a for-loop is about 2 orders of magnitude slower than the first in calculating the sums. This doesn't include the copy but, as I said before, it turns out the copy isn't that big of a deal time-wise.

Another potential point of confusion might be someone who wants to get the value of the last node.
```python
>>> z = np.arange(12, dtype=float)
>>> grid.at_node['myvals'] = z
>>> grid.at_node['myvals'][-1]
0.
```
I would think they are going to expect to get back the same array they put in.

Perhaps we could have `grid.at_node['myvals']` by default return the view that doesn't have the extra value. There could be some other way for those folks who knew what they were doing to get the view of the array with the extra value. I can't think of what that way would be, though.

One possibility ("x" for extended),
```
>>> grid.at_node['myvals']
array([0, 1, 2, 3])
>>> grid.at_node['myvals'].x
array([0, 1, 2, 3, 0])
```

Maybe you're right. Maybe it is too confusing.

**Comment by mcflugen** [comment]
And, yes @gregtucker, I will have a look! Sorry. We highjacked your thread.

**Comment by SiccarPoint** [comment]
Yeah. It feels like it should work, but once you start chasing down the possibilities of how to make it work, you end up in the weeds really fast. I think you’ve put your finder on the key difficulty. Probably just leave it as it is, then, and make sure the docstrings on any function that uses -1s is *super* clear on what this means. Though that said, I still kind of like the idea of building effectively a “child safe lock” onto the method arguments:

>>> mg.node_neighbors(are_you_an_idiot_who_doesnt_understand_this_has_minusones_in_it=True)

Kidding aside, this would look like

>>>mg.node_neighbors(no_neighbor=BAD_INDEX_VALUE)

…where the other option is -1. We’ll always be calling the -1 option, but this will default to BAD_INDEX_VALUE, triggering the user to realise there’s an issue and actually check the documentation… where they will see the -1 info.

**Comment by mcflugen** [comment]
Sorry. I'm kind of stuck on this idea now. I just pushed a small change to master that adds a new function, `extend_array`. I don't know, it might be useful for the things we're talking about.
```python
>>> from landlab.core.utils import extend_array
>>> import numpy as np
>>> arr = extend_array(np.arange(4).reshape((2, 2)))
>>> arr
array([[0, 1],
       [2, 3]])
>>> arr.x
array([0, 1, 2, 3, 0])
```
So now we have an array that looks like the original and shares its data. However, it has a couple differences. Its data buffer has been extended by one element and that element is filled with a 0 (use the *fill* keyword to change the fill value). You can access this extended array with the new `x` attribute.

**Comment by SiccarPoint** [comment]
“ext” might be clearer than “x”? x suggests an x-coord.

But awesome!

**Comment by SiccarPoint** [comment]
Also, it might be cool to have fields be created like this by default? Then we can just tell users it’s field functionality they can use if they want, and apart from a bit of overhead on creation, it shouldn’t make much difference otherwise…?

**Comment by mcflugen** [comment]
We can totally do that. There would only be additional overhead if they add a previously-created array as a field. If they use the field construction methods (like `grid.add_zeros`) which I think is the most common way, there's no extra overhead.

**Comment by gregtucker** [comment]
Actually we already use the array-extension trick, inefficiently, in calculate_flux_divergence. That's WHY the node-link matrices use -1 instead of BIV. But at the moment we use copying: every time user sends a link array of data, we copy it into a NUM_LINKS+1 array and then do the flux divergence calculation by summing fluxes in each row of the matrix.

This gets to the heart of my project to revisit node-link data structures! (to which re-ordering links is just the prelude...) I don't love the current design. I want to replace the following four current arrays:
node_inlink_matrix, node_outlink_matrix, node_active_inlink_matrix, node_active_outlink_matrix
Each of these is 2xN for raster, 6xN for hex, and MAX_NBRSxN for general Voronoi and Radial.

There are a couple of possible alternatives I want to explore. One, which I've mentioned before, is the following:
links_at_node: 4xN for raster, 6xN for hex, etc. Code -1 means "no such link", as before.
link_dirs_at_node: same size; coded 1 if node is link's head (i.e., an inlink), -1 if tail (outlink), and 0 if no such link. THAT'S KEY!
active_link_dirs_at_node: same as above but with 0 for inactive links.
With this method, you don't need to extend link arrays to do flux calculations. Flux divergence, for a raster, becomes something like the following:

for i in range(4):
    fd += flux[links_at_node[i,:]] * active_link_dirs_at_node[i,:]

Key problematic assumption: link_dirs_at_node[:,-1] = -1 ... in other words, the last node in the array NEVER has active links. In a raster this should always be true. Not totally sure about others. But if it IS guaranteed true, then the flux for any non-existent or inactive link is always zero.

Another possibility is MASKED ARRAYS (which I still don't fully understand)

Another is Eric's suggestion of flat arrays. This might look something like:

link_at_node_index: Nx1, containing the index at which links (and neighbor nodes, etc) for a given node begin.
link_at_node: (>N)x1, generally slightly less than 4xN for a raster, 6xN for hex, etc.
Problem: not obvious whether there's a "numpyish" solution. Fancy indexing is one possibility.

A variant on this: as above, but use looping and Cython.

Anyway, for now, I will change doctests on the mappers to the behavior that they actually express, rather than what we'd ideally hope.
